### PR TITLE
feat(op-challenger): Record Cannon Execution Time

### PR DESCRIPTION
**Pull request by refcell**
**Description**

Adds a prometheus histogram to record cannon execution time by propagating
the Metricer down through the challenger to the cannon trace provider.

The executor accepts a minimal metrics interface containing the single
cannon execution time record method.

**Metadata**

Fixes CLI-4393

### CODE REVIEW

**Comment by refcell** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #7078** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/7078" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/7078?utm_source=stack-comment).

File: op-challenger/game/fault/trace/cannon/executor.go [hunk 1]
```diff
@@ -131,7 +134,13 @@ func runCmd(ctx context.Context, l log.Logger, binary string, args ...string) er
...
 	cmd.Stdout = stdOut
 	cmd.Stderr = stdErr
-	return cmd.Run()
+	start := time.Now()
+	err := cmd.Run()
+	if err != nil {
+		execTime := time.Since(start).Seconds()
+		m.RecordCannonExecutionTime(execTime)
+	}
+	return err
```
Comment (ajsutton) [inline op-challenger/game/fault/trace/cannon/executor.go]: I'd suggest tracking this metric outside this `runCmd` so we can keep this really minimally focussed on the difficult-to-test actual execution of a process.  Effectively the time it takes for `runCmd` to execute is the cannon execution time as far as op-challenger is concerned.

File: op-challenger/game/fault/trace/cannon/executor_test.go [hunk 2]
```diff
@@ -210,3 +212,7 @@ func TestFindStartingSnapshot(t *testing.T) {
 		require.Equal(t, filepath.Join(dir, "100.json.gz"), snapshot)
 	})
 }
+
+type mockMetricer struct{}
```
Comment (ajsutton) [inline op-challenger/game/fault/trace/cannon/executor_test.go]: Should be able to just use `NoopMetrics` instead.

File: op-challenger/game/monitor_test.go [hunk 3]
```diff
@@ -99,11 +100,20 @@ func setupMonitorTest(t *testing.T, allowedGames []common.Address) (*gameMonitor
...
 	return monitor, source, sched
 }
 
+type mockMetricer struct {
+	txmetrics.NoopTxMetrics
+}
+
+func (m *mockMetricer) RecordInfo(version string)           {}
+func (m *mockMetricer) RecordUp()                           {}
+func (m *mockMetricer) RecordCannonExecutionTime(t float64) {}
```
Comment (ajsutton) [inline op-challenger/game/monitor_test.go]: NoopMetrics. :)

File: op-challenger/metrics/metrics.go [hunk 4]
```diff
@@ -58,6 +62,12 @@ func NewMetrics() *Metrics {
 			Name:      "up",
 			Help:      "1 if the op-challenger has finished starting up",
 		}),
+		cannonExecutionTime: factory.NewHistogram(prometheus.HistogramOpts{
+			Namespace: Namespace,
+			Name:      "cannon_execution_time",
+			Help:      "Time (in seconds) to execute cannon",
+			Buckets:   append([]float64{1.0, 10.0}, prometheus.LinearBuckets(30.0, 30.0, 14)...),
```
Comment (ajsutton) [inline op-challenger/metrics/metrics.go]: If I understand this correctly it's only tracking up to 7 minutes but the initial run of cannon is expected to take between 1 and 2 hours. I suspect we want to use exponential buckets.

File: op-e2e/e2eutils/disputegame/metricer.go [hunk 5]
```diff
@@ -0,0 +1,13 @@
+package disputegame
+
+import (
+	"github.com/ethereum-optimism/optimism/op-service/txmgr/metrics"
+)
+
+type mockMetricer struct {
```
Comment (ajsutton) [inline op-e2e/e2eutils/disputegame/metricer.go]: This is just NoopMetrics. :)

File: op-challenger/metrics/metrics.go (hunk 4 above)
Comment (refcell) [reply op-challenger/metrics/metrics.go]: Updated to use exponential buckets which should cover a few hours if i did my math right

File: op-challenger/metrics/metrics.go [hunk 6]
```diff
@@ -58,6 +62,12 @@ func NewMetrics() *Metrics {
 			Name:      "up",
 			Help:      "1 if the op-challenger has finished starting up",
 		}),
+		cannonExecutionTime: factory.NewHistogram(prometheus.HistogramOpts{
+			Namespace: Namespace,
+			Name:      "cannon_execution_time",
+			Help:      "Time (in seconds) to execute cannon",
+			Buckets:   append([]float64{1.0, 10.0}, prometheus.ExponentialBuckets(30.0, 2.0, 14)...),
```
Comment (ajsutton) [inline op-challenger/metrics/metrics.go:83]: What about just using:
```suggestion
			Buckets:   append([]float64{1.0, 10.0}, prometheus.ExponentialBucketsRange(30.0, 60.0 * 3.0, 14)...),
```

Then we're defining the range we want and the number of buckets and letting prometheus do the math. :)

**Review by ajsutton — APPROVED** [review]
LGTM. Just the one suggestion for how to define the buckets.

File: op-challenger/metrics/metrics.go (hunk 6 above)
Comment (refcell) [reply op-challenger/metrics/metrics.go:83]: That looks good to me - can always lower or raise the max when we start testing metrics :)

Comment (ajsutton) [reply op-challenger/metrics/metrics.go:83]: Yeah 3 hours is pretty arbitrary but I think we'd want to find a faster CPU if cannon is taking longer than that...

**Pull request merged**